examples/hybrid_jobs/6_QNSPSA_optimizer_with_embedded_simulator/qnspsa_with_embedded_simulator.ipynb

```diff
@@ -0,0 +1,1360 @@
+      "        op_forward = self.__get_operations(cost, params1)\n",
+      "        op_inv = self.__get_operations(cost, params2)\n",
+      "\n",
+      "        with qml.tape.QuantumTape() as tape:\n",
+      "            for op in op_forward:\n",
+      "                qml.apply(op)\n",
+      "            for op in reversed(op_inv):\n",
+      "                op.adjoint()\n",
+      "            #TODO: update this line in PL paper project repo.\n",
+      "            qml.probs(wires=cost.tape.wires.labels)\n",
+      "        return tape\n",
+      "               \n",
+      "    def __get_operations(self, cost, params):\n",
+      "        cost.construct([params], {})\n",
+      "        return cost.tape.operations     \n",
+      "               \n",
+      "    def __get_tensor_moving_avg(self, metric_tensor):\n",
+      "        if self.metric_tensor is None:\n",
+      "            self.metric_tensor = np.identity(metric_tensor.shape[0])        \n",
+      "        return (self.k / (self.k + 1) * self.metric_tensor + \n",
+      "                1 / (self.k + 1) * metric_tensor)\n",
+      "      \n",
+      "    def __regularize_tensor(self, metric_tensor):\n",
+      "        tensor_reg = np.real(sqrtm(np.matmul(metric_tensor, metric_tensor)))\n",
+      "        return ((tensor_reg + self.reg * np.identity(metric_tensor.shape[0])) /\n",
+      "                (1 + self.reg))  \n",
+      "    \n",
+      "    def __apply_blocking(self, cost, params_curr, params_next):\n",
+      "        cost.construct([params_curr], {})\n",
+      "        tape_loss_curr = cost.tape.copy(copy_operations=True)\n",
+      "        cost.construct([params_next], {})\n",
+      "        tape_loss_next = cost.tape.copy(copy_operations=True)        \n",
+      "        \n",
+      "        loss_curr, loss_next = qml.execute(\n",
+      "            [tape_loss_curr, tape_loss_next], \n",
+      "            cost.device, None)\n",
+      "        #self.k has been updated earlier\n",
+      "        ind = (self.k - 2) % self.history_length\n",
+      "        self.last_n_steps[ind] = loss_curr\n",
+      "        \n",
+      "        tol = 2 * self.last_n_steps.std() if self.k > self.history_length \\\n",
+      "            else 2 * self.last_n_steps[:self.k - 1].std()  \n",
+      "        \n",
+      "        if loss_curr + tol < loss_next:\n",
+      "            params_next = params_curr\n",
+      "        return params_next, loss_curr"
+     ]
+    }
+   ],
+   "source": [
+    "!cat source_scripts/QNSPSA.py"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "448931d5",
+   "metadata": {},
+   "source": [
+    "To confirm the optimizer works, here I choose to work with a QAOA example on a toy maximum cut problem [6]. \n",
+    "\n",
+    "The maximum cut is the problem to find a cut onto a given graph that leads to the most edges between the two resulting subsets of the original graph. The next cell sets up the max cut problem. "
```

Comment:
Although the problem is called max _cut_, I find it confusing to actually speak about cutting a graph, which trying to maximize edge numbers. Let's phrase it in terms of partitioning: "The maximum cut problem is partitioning the nodes of a graph into two subsets such that the number of edges going between the subsets is maximized."